### PR TITLE
Fix: Underselection test bug

### PR DESCRIPTION
**Pull request by petrroll**
+ Fix for #38708 
+ Enabled interleaving of {| and |] in tests (required for the issue regression tests)
+ Added general tests for IsUnderselected + machinery.

### CODE REVIEW

**Comment by petrroll** [comment]
Not sure I'm targeting the correct branch, if possible this should get fixed before next 16.4 preview. It also needs to flow to master but I assume that's still automatic (and this branch isn't abandoned).

File: src/Features/Core/Portable/CodeRefactoringHelpers.cs [hunk 1]
```diff
@@ -30,45 +30,66 @@ internal static class CodeRefactoringHelpers
         /// it is the smallest that can be retrieved.
         /// </para>
         /// <para>
+        /// When selection doesn't contain intersect the node in any way it's not considered to not to be underselected.
```
Comment (miloush) [inline src/Features/Core/Portable/CodeRefactoringHelpers.cs]: forgotten "contain" and too many negatives at the end

File: src/Features/Core/Portable/CodeRefactoringHelpers.cs [hunk 2]
```diff
@@ -30,45 +30,66 @@ internal static class CodeRefactoringHelpers
...
+            // RATIONALE: If there's no intersection then we got the node in some other way, e.g. 
+            // extracting it after user selected `;` at the end of an expression statement 
+            // `foo()[|;|]` for `foo()` node.
+            if (!node.FullSpan.OverlapsWith(selection))
             {
                 return false;
             }
 
+            // Only precisely one token of the node is selected -> treat is as empty selection -> not 
+            // under-selected. The rationale is that if a only one Token is selected then the selection
```
Comment (miloush) [inline src/Features/Core/Portable/CodeRefactoringHelpers.cs]: "underselected" for consistency and forgotten "a"

**Comment by mavasani** [comment]
This branch is no longer maintained. You can retarget to master. Thanks!

**Comment by jinujoseph** [comment]
@petrroll thanks for the fix, would you pls retarget this to master pls

File: src/Features/Core/Portable/CodeRefactoringHelpers.cs [hunk 3]
```diff
@@ -30,45 +30,66 @@ internal static class CodeRefactoringHelpers
...
         public static bool IsNodeUnderselected(SyntaxNode? node, TextSpan selection)
         {
+            // Selection is null -> it's always underselected
+            // REASON: Easier API use -> underselected node, don't work on it further
             if (node == null)
             {
                 return true;
             }
 
+            // Selection or node is empty -> can't be uderselected
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactoringHelpers.cs]: Uder

File: src/Features/Core/Portable/CodeRefactoringHelpers.cs [hunk 4]
```diff
@@ -30,45 +30,66 @@ internal static class CodeRefactoringHelpers
...
+            }
+
             var beginningNode = node.FindToken(node.Span.Start).Parent;
             var endNode = node.FindToken(node.Span.End - 1).Parent;
 
             // Node is underselected if either the first (lowest) child doesn't contain start of selection
             // of the last child doesn't intersect with the end.
+
+            // Node is underselected if either the first (lowest) child ends before the selection has started
+            // of the last child starts before the selection ends (i.e. one of them is completely on the outside of selection).
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactoringHelpers.cs]: Of -> or

File: src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs [hunk 5]
```diff
@@ -58,11 +58,12 @@ public static class MarkupTestFile
 
             // A stack of span starts along with their associated annotation name.  [||] spans simply
             // have empty string for their annotation name.
-            var spanStartStack = new Stack<Tuple<int, string>>();
+            var spanStartStack = new Stack<(int, string)>();
+            var namedSpanStartStack = new Stack<(int, string)>();
```
Comment (CyrusNajmabadi) [inline src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs]: Why this change?

Comment (petrroll) [reply src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs]: Less allocation, old class-based tuples are slowly being replaced where possible.

**Comment by petrroll** [comment]
Feedback addressed + cleaned up Tuple usage.

File: src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs (hunk 5 above)
Comment (petrroll) [reply src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs]: Ability to use names instead of Element1 and Element2 (see new commit).

Comment (CyrusNajmabadi) [reply src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs]: Sorry, i meant specifically: why the switch to two stacks?

File: src/Features/Core/Portable/CodeRefactoringHelpers.cs [hunk 6]
```diff
@@ -30,45 +30,66 @@ internal static class CodeRefactoringHelpers
...
 
             // Node is underselected if either the first (lowest) child doesn't contain start of selection
             // of the last child doesn't intersect with the end.
+
+            // Node is underselected if either the first (lowest) child ends before the selection has started
+            // or the last child starts before the selection ends (i.e. one of them is completely on the outside of selection).
             // It's a crude heuristic but it allows omitting parts of nodes or trivial tokens from the beginning/end 
             // but fires up e.g.: `1 + [|2 + 3|]`.
-            return !beginningNode.Span.IntersectsWith(selection.Start) || !endNode.Span.IntersectsWith(selection.End - 1);
+            return beginningNode.Span.End <= selection.Start || endNode.Span.Start > selection.End - 1;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactoringHelpers.cs]: instead of `> .End - 1` why not `>= End`?

File: src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs (hunk 5 above)
Comment (petrroll) [reply src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs]: So that the named & not named can interleave each other. E.g.:
```
        //[|>
        var str = {|result:"" <|] aaa""|};
```

Previously only nesting was allowed

File: src/Features/Core/Portable/CodeRefactoringHelpers.cs (hunk 6 above)
Comment (petrroll) [reply src/Features/Core/Portable/CodeRefactoringHelpers.cs]: Artifact of change. That said, seems ?more explicit? to me?

File: src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Test/Utilities/Portable/MarkedSource/MarkupTestFile.cs]: terrifying!  but ok :D

File: src/Features/Core/Portable/CodeRefactoringHelpers.cs (hunk 6 above)
Comment (petrroll) [reply src/Features/Core/Portable/CodeRefactoringHelpers.cs]: Wouldn't fix it normally but want to retrigger tests because of flaky Vb test anyway. 
(Not having permissions to just rerun tests is serious annoyance).

**Comment by petrroll** [comment]
Please review @mavasani

**Comment by mavasani** [comment]
Thanks @petrroll!
